src/cpu/x64/matmul_inner_product.cpp

```diff
@@ -0,0 +1,271 @@
+        { BA16a64b4a, {AB16b64a4b, AcB16b64a4b, AcdB16b64a4b, AcdeB16b64a4b}}};
+    // clang-format on
+
+    auto mm_wei_tag = format_tag::undef;
+    if (weights_md()->format_kind != format_kind::any) {
+        for (const auto &v : mm_wei_to_ip_wei) {
+            if (memory_desc_matches_tag(
+                        *weights_md(), v.second[weights_md()->ndims - 2])) {
+                mm_wei_tag = v.first;
+                // Check if the user-provided blocked layout can be handled.
+                const bool has_spatial = KD() + KH() + KW() > 3;
+                const int k_blk = get_k_blk(mm_wei_tag);
+                if (weights_md()->ndims > 2 && has_spatial && k_blk > 0
+                        && IC() % k_blk != 0)
+                    return status::unimplemented;
+                break;
+            }
+        }
+    } else {
+        mm_wei_tag = format_tag::any;
+    }
+    if (mm_wei_tag == format_tag::undef) return status::unimplemented;
+
+    memory_desc_t mm_src_md {};
+    memory_desc_t mm_wei_md {};
+    memory_desc_t mm_dst_md {};
+
+    if (bias_md_.format_kind == format_kind::any)
+        CHECK(memory_desc_init_by_tag(bias_md_, x));
+
+    CHECK(init_matmul_md(mm_src_md, *src_md(), format_tag::ab));
+    CHECK(init_matmul_md(mm_wei_md, *weights_md(), mm_wei_tag, true));
+    CHECK(init_matmul_md(mm_dst_md, *dst_md(), format_tag::ab));
+
+    const auto src_tag = utils::pick(src_md()->ndims - 2, ab, acb, acdb, acdeb);
+    if (src_md()->format_kind == format_kind::any)
+        CHECK(memory_desc_init_by_tag(src_md_, src_md_.ndims, src_md_.dims,
+                src_md_.data_type, src_tag));
+    else if (!memory_desc_matches_tag(*src_md(), src_tag))
+        return status::unimplemented;
+
+    const auto dst_tag = ab;
+    if (dst_md()->format_kind == format_kind::any)
+        CHECK(memory_desc_init_by_tag(dst_md_, dst_md_.ndims, dst_md_.dims,
+                dst_md_.data_type, dst_tag));
+    else if (!memory_desc_matches_tag(*dst_md(), dst_tag))
+        return status::unimplemented;
+
+    VDISPATCH_INNER_PRODUCT_SC(
+            attr_.set_default_formats(dst_md(0)), VERBOSE_UNSUPPORTED_POSTOP);
+
+    primitive_attr_t matmul_attr = *attr();
+    const auto wei_mask = matmul_attr.scales_.get(DNNL_ARG_WEIGHTS).mask_;
+    if (wei_mask == 1)
+        VDISPATCH_INNER_PRODUCT_SC(matmul_attr.scales_.set(DNNL_ARG_WEIGHTS,
+                                           1 << (mm_wei_md.ndims - 1)),
+                VERBOSE_UNSUPPORTED_ATTR);
+    else if (wei_mask != 0)
+        return status::unimplemented;
+
+    CHECK(create_matmul_pd(matmul_pd_, engine, &mm_src_md, &mm_wei_md,
+            &mm_dst_md, with_bias() ? weights_md(1) : nullptr, &matmul_attr));
+
+    // Fallback to a generic GEMM-based Inner Product is preferred rather
+    // than using a reference or GEMM-based MatMul implementations here.
+    if (std::string(matmul_pd_->name()).find("ref") != std::string::npos)
```

Comment:
Minor: maybe use the inversion to include brg_matmul only to be more precautious?

Reply:
Replaced the condition